### PR TITLE
fix: select node with most blocks for stream validation in restart tests

### LINKED ISSUE #23858: Test flake: StreamValidationTest > streamsAreValid()
```
> Task :test-clients:testSubprocess

  com.hedera.services.bdd.suites.validation.StreamValidationTest > streamsAreValid() ✘ as written (18.5s)

    java.lang.AssertionError: Block stream validation failed:
      - Field blockFooter is null
      - Block must end with a block proof
```

https://github.com/hiero-ledger/hiero-consensus-node/actions/runs/22391134658/job/64814349644?pr=23831
https://github.com/hiero-ledger/hiero-consensus-node/actions/runs/22661288469/job/65682427536?pr=23927

Find the root cause and fix the issue.

### PR DESCRIPTION
**Pull request by AlexKehayov**
**Description**:
- In restart tests, nodes stopped mid-lifecycle leave behind incomplete block files (.pnd files or .blk files whose async proof was never written), causing "blockFooter is null" and "Block must end with a block proof" validation failures                                                                                                                                                                                                         
- The original code took the first non-empty node's block set and passed it directly to validators, which would fail on any incomplete block
- Now selects the node directory with the most block files for validation, instead of stopping at the first non-empty one                                                                                                                                                                                         
- Substitutes incomplete blocks (missing footer/proof) from other node directories via direct path resolution
- Hardens StateChangesValidator and BlockContentsValidator to tolerate incomplete blocks in the middle of the block list, which occur when all nodes write a block before the async proof arrives during restarts 

This is a follow-up to #24013 which fixed a related issue where blocks spread across multiple block-X.Y.Z account-ID subdirectories were not all discovered.

Fixes #23858

### CODE REVIEW

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/StreamValidationOp.java [hunk 1]
```diff
@@ -188,25 +192,109 @@ protected boolean submitOp(@NonNull final HapiSpec spec) throws Throwable {
...
-                // We will try to read the next node's streams
+                // We will try the next node's directory
             }
-            if (blocks != null && !blocks.isEmpty()) {
-                break;
+        }
+        if (bestDir == null) {
+            return Optional.empty();
+        }
+        final var best = BLOCK_STREAM_ACCESS.readBlocks(bestDir);
```
Comment (petreze) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/StreamValidationOp.java]: This will be better to be wrapped up in a `try-catch` block

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/validators/block/StateChangesValidator.java [hunk 2]
```diff
@@ -494,15 +501,20 @@ public void validateBlocks(@NonNull final List<Block> blocks) {
...
+                            && nextBlockItems.get(nextBlockFooterIndex).hasBlockFooter()) {
+                        previousBlockHash = nextBlockItems
+                                .get(nextBlockFooterIndex)
+                                .blockFooterOrThrow()
+                                .previousBlockRootHash();
+                    }
                 }
 
                 incrementalBlockHashes.addNodeByHash(previousBlockHash.toByteArray());
+            } else if (i <= lastVerifiableIndex) {
```
Comment (petreze) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/validators/block/StateChangesValidator.java:523]: Is this the case where all the nodes have the same non-finalized block file? If, so shouldn't we fail the StateChangesValidator here, instead of just logging?

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/StreamValidationOp.java [hunk 3]
```diff
@@ -188,25 +192,109 @@ protected boolean submitOp(@NonNull final HapiSpec spec) throws Throwable {
...
+            result.add(replacement != null ? replacement : block);
+        }
+        return Optional.of(result);
+    }
+
+    /**
+     * Collects block file paths from all node directories except the best one, keyed by block
+     * number. Used as a fallback when the primary node has incomplete blocks that need substitution.
+     */
+    private static TreeMap<Long, List<Path>> collectFallbackBlockFilePaths(
```
Comment (Grigorov-Georgi) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/StreamValidationOp.java]: Do we really need TreeMap? I don't see any iterations requiring sorted keys. So returning Map interface + using HashMap for the impl seems like a better fit for doing only lookups by key.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/validators/block/StateChangesValidator.java (hunk 2 above)
Comment (AlexKehayov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/validators/block/StateChangesValidator.java:523]: If we fail here, the flake will persist. Incomplete blocks in the middle are an expected edge case during restarts. The state changes are still applied and validated. We only skip the proof/footer check for that one block. Should we think about something like adding a threshold for failing if more than X blocks are skipped?

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/StreamValidationOp.java (hunk 1 above)
Comment (AlexKehayov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/StreamValidationOp.java]: You are right, nice catch, thanks

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/StreamValidationOp.java (hunk 3 above)
Comment (AlexKehayov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/utilops/streams/StreamValidationOp.java]: That's true, it was a leftover from a previous version of the fix, nice catch, thank you

**Review by Grigorov-Georgi — APPROVED** [review]

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/validators/block/StateChangesValidator.java [hunk 4]
```diff
@@ -443,9 +443,16 @@ public void validateBlocks(@NonNull final List<Block> blocks) {
...
+            // when nodes are restarted and all nodes wrote the block before the async proof arrived
+            final long blockNumber =
+                    block.items().getFirst().blockHeaderOrThrow().number();
+            final boolean blockHasProof = block.items().getLast().hasBlockProof();
+            if (i <= lastVerifiableIndex && blockHasProof) {
                 final var footer = block.items().get(block.items().size() - 2);
-                assertTrue(footer.hasBlockFooter());
+                assertTrue(
+                        footer.hasBlockFooter(),
+                        "Field blockFooter is null for block #" + blockNumber + " at index " + i);
```
Comment (ibankov) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/validators/block/StateChangesValidator.java:448]: From Codex:
In StateChangesValidator.java:451 the new blockHasProof guard skips verification for incomplete blocks, but in the skip path at StateChangesValidator.java:516 previousBlockHash is not advanced. The next complete block is still checked against this stale hash at line 460, so a sequence like complete -> incomplete -> complete still fails. That contradicts the PR goal of tolerating incomplete blocks in the middle.
Have you considered this?

Comment (AlexKehayov) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/validators/block/StateChangesValidator.java:448]: Makes sense, nice catch. Adding it in the next commit.
